### PR TITLE
updates to hubs document

### LINKED ISSUE #9497: Detect intended/unintended client disconnect
It is not clear from the documentation if there is any way to distinguish if a client was disconnected from a hub due to technical problems (connectivity issues, application exception, ...) or if it was intended (client closed window, ...).

Looking at the signature, `OnDisconnectedAsync(Exception exception)` I am lead to believe that `exception` will be null when all is fine, and always set to something if the disconnect was unintentional. Is this correct?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1387f556-b4b7-70f9-5bd0-f0bccc597689
* Version Independent ID: fb016cdc-d281-6b5b-7466-2c718e7274b4
* Content: [Use hubs in ASP.NET Core SignalR](https://docs.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-2.1#handle-events-for-a-connection)
* Content Source: [aspnetcore/signalr/hubs.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/signalr/hubs.md)
* Product: **aspnet-core**
* GitHub Login: @tdykstra
* Microsoft Alias: **tdykstra**

### PR DESCRIPTION
**Pull request by anurse**
fixes #9497 as well

Some updates that came up as I was writing #9455 but didn't fit there.

[Internal Review Link](https://review.docs.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-2.1&branch=pr-en-us-9501)

### CODE REVIEW

File: aspnetcore/signalr/hubs.md [hunk 1]
```diff
@@ -91,9 +91,13 @@ Each property or method in the preceding tables returns an object with a `SendAs
 
 ## Send messages to clients
 
-To make calls to specific clients, use the properties of the `Clients` object. In the following example, the `SendMessageToCaller` method demonstrates sending a message to the connection that invoked the hub method. The `SendMessageToGroups` method sends a message to the groups stored in a `List` named `groups`.
+To make calls to specific clients, use the properties of the `Clients` object. In the following example, there are three Hub methods:
 
-[!code-csharp[Send messages](hubs/sample/hubs/chathub.cs?range=15-24)]
+* `SendMessage` sends a message to all connected clients, using `Clients.All`.
+* `SendMessageToCaller` sends a message back to caller, using `Clients.Caller`.
+* `SendMessageToGroups` sends a message to all clients in the `SignalR Users` group.
```
Comment (BrennanConroy) [inline aspnetcore/signalr/hubs.md]:
```suggestion
* `SendMessageToGroups` sends a message to all clients in the `SignalR Users` group, using `Clients.Group`.
```

File: aspnetcore/signalr/hubs.md [hunk 2]
```diff
@@ -111,18 +115,43 @@ Using `Hub<IChatClient>` enables compile-time checking of the client methods. Th
...
+
+```
+Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'MethodName' on the server.
+```
+
+Unexpected exceptions often contain sensitive information, such as the name of a database server in an exception triggered when the database connection fails. SignalR doesn't expose these detailed error messages by default as a security measure. See the [Security considerations article](xref:signalr/security#exceptions) for more information on why exception details are suppressed.
+
+If you have an exceptional condition you *do* want to propagate to the client, you can use the `HubException` class. If you throw a `HubException` from your hub method, SignalR **will** send the entire message to the client, unmodified.
+
+[!code-csharp[ThrowHubException](hubs/sample/hubs/chathub.cs?name=ThrowHubException&highlight=3)]
```
Comment (BrennanConroy) [inline aspnetcore/signalr/hubs.md]: code snippet isn't regioning correctly

**Comment by anurse** [comment]
:up: :date: I had to move one of the snippets to be an inline snippet because when I added the `#region` markers it made the "summary" sample at the top of the page ugly. I think the regions are really useful and that initial sample is just there to show a very basic hub, so I think moving it inline works well.

Also addressed @tdykstra 's feedback.

File: aspnetcore/signalr/hubs.md [hunk 3]
```diff
@@ -5,7 +5,7 @@ description: Learn how to use hubs in ASP.NET Core SignalR.
 monikerRange: '>= aspnetcore-2.1'
 ms.author: tdykstra
 ms.custom: mvc
-ms.date: 09/12/2018
+ms.date: 11/7/2018
```
Comment (scottaddie) [inline aspnetcore/signalr/hubs.md]:
```suggestion
ms.date: 11/07/2018
```

File: aspnetcore/signalr/hubs.md [hunk 4]
```diff
@@ -111,18 +123,43 @@ Using `Hub<IChatClient>` enables compile-time checking of the client methods. Th
...
+```
+
+Unexpected exceptions often contain sensitive information, such as the name of a database server in an exception triggered when the database connection fails. SignalR doesn't expose these detailed error messages by default as a security measure. See the [Security considerations article](xref:signalr/security#exceptions) for more information on why exception details are suppressed.
+
+If you have an exceptional condition you *do* want to propagate to the client, you can use the `HubException` class. If you throw a `HubException` from your hub method, SignalR **will** send the entire message to the client, unmodified.
+
+[!code-csharp[ThrowHubException](hubs/sample/hubs/chathub.cs?name=ThrowHubException&highlight=3)]
+
+> [!NOTE]
+> SignalR only sends the `Message` property of the exception to the client. The stack trace and other properties on the exception are not available to the client.
```
Comment (scottaddie) [inline aspnetcore/signalr/hubs.md]: are not --> aren't

File: aspnetcore/signalr/hubs/sample/Hubs/ChatHub.cs [hunk 5]
```diff
@@ -7,32 +7,52 @@ namespace SignalRChat.Hubs
...
+        #region HubMethodName
+        [HubMethodName("SendMessageToUser")]
+        public Task DirectMessage(string user, string message)
+        {
+            return Clients.User(user).SendAsync("ReceiveMessage", message);
+        }
+        #endregion
+
+        #region ThrowHubException
+        public async Task ThrowException()
```
Comment (BrennanConroy) [inline aspnetcore/signalr/hubs/sample/Hubs/ChatHub.cs]: nit: remove async, this wouldn't compile as is.

Comment (anurse) [reply aspnetcore/signalr/hubs/sample/Hubs/ChatHub.cs]: It'll compile, but with a warning (CS1998: This async method lacks 'await' operators and will run synchronously...). I originally intended to model an "async" exception here and using `async` ensures that a state-machine is used. We handle synchronous exceptions too though so I think I will switch it up to drop `async`.

**Comment by anurse** [comment]
Feedback should be all addressed. @scottaddie, @tdykstra let me know if there's anything else or if this is ready to go! Thanks for the feedback all!

**Comment by tdykstra** [comment]
LGTM

**Comment by scottaddie** [comment]
Thank you for another awesome PR, @anurse!

**Pull request merged**